### DISCUSSION TITLE
Parametrized surface

### DISCUSSION THREAD

**Discussion by theottm** [kaveh808/kons-9#251] [Ideas]
I see three elements to describe this:
- a function from the plane to 3D space
- a domain on which the function is defined
- a sampling method for the domain

The domain is just a rectangular grid. But like the uv-mesh it needs additional infos to describe its topology ([how the sides of the rectangle are glued together](https://en.wikipedia.org/wiki/Surface_(topology)#Construction_from_polygons))

So with the three building blocks at hand one can describe arbitrary surfaces it seems.

**Comment by kaveh808** [comment]
That makes sense. The sampling method would lead to either mesh tessellation or direct GPU sampling?

What concrete surfaces were you thinking of implementing? Spline surfaces, NURBS?

**Reply by theottm** [reply]
You are right "sampling" is an ambiguous term. As you said one wants 1. to tesselate the domain first to get a mesh and then 2. map all the vertices to coordinates.

It seems to me like these two operations are actually rather general and not only appear in the context of surfaces. So both might relate to existing parts of kons-9 and might also be of interest to do other things.

For 1:
I remember having seen a form of mesh refinement in the demos, maybe we can elaborate on that for tesselation. I'm not that aware of tesselation methods and related considerations beyond deviding a big rectangle into many smaller rectangles. I would focus on that first, fancier stuff can come later on.

For 2:
For the application of a map to vertices of a mesh, we of course already have linear transformations. Not sure if we can build on this though. As you said I think it would be nice to apply the transformations through a shader (one more reason to have the CL-GLSL DSL and the modern OpenGL pipeline). We might also want to compute the values on a CPU for some cases (some function are defined iteratively, one could also want to get coordinates from an external process). This can actually be used as a general mean of applying a transformation to a mesh (for example applying an inversion in a sphere).

I have near zero experience with spline surfaces and NURBS, even if they look like very handy modeling tools. The kind of surfaces I generally visualize are rather theoretical, like surfaces of revolution, developable surfaces, some embedded tori, klein bottles, Moebius bands, Dupin cyclides, quadrics, etc.

**Comment by kaveh808** [comment]
The `subdiv-mesh` subdivision uses Catmull-Rom refinement, which divides each n-gon into n quads.

We also have superquadrics as a shape class.

I'm all for generalizing techniques in this area and denoting the relationships between the concrete classes.
